Objects/call.c

```diff
@@ -309,6 +309,30 @@ _PyObject_FastCall(PyObject *func, PyObject *const *args, Py_ssize_t nargs)
 }
 
 
+static PyObject*
+_new_dict_if_nonempty(PyObject *mapping)
+{
+    PyObject *newdict = NULL;
+
+    if (mapping != NULL) {
+        Py_ssize_t kwlen = PyDict_Size(mapping);
```

Comment:
This fails with a `PyErr_BadInternalCall` if mapping is not a dict (and cannot fail if it is).

Reply:
`PyObject_Call` has an assert that verifies `assert(kwargs == NULL || PyDict_Check(kwargs));`, so the mapping is definitely expected to be a `dict` or subclass thereof by the time we reach this line. I'll admit the error-check is unnecessary at the present time, since `PyDict_Size` can't fail, but it felt weird not checking when the function might, at some point in the future, have other situations in which it raises an error.

If you don't like it as is, I'll happily change:

            if (newdict == NULL) {
                return NULL;
            }

to a simple:

            assert(newdict != NULL);

that verifies behavior in debug mode and imposes no costs on normal interpreter builds. That seem reasonable?

Reply:
Sure. Could you add an `assert(PyDict_Check(mapping))` as well.

Reply:
That seems kind of redundant with `PyObject_Call`'s check, but I suppose if the function ever gets used elsewhere it'll be useful. Will do.